src/webui/api/torrentscontroller.cpp

```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
     const QString hash = params()["hash"];
 
     BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
-    if (torrent) {
-        QList<BitTorrent::TrackerEntry> trackers;
-        foreach (QString url, params()["urls"].split('\n')) {
-            url = url.trimmed();
-            if (!url.isEmpty())
-                trackers << url;
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QList<BitTorrent::TrackerEntry> trackers;
+    for (QString urlStr : params()["urls"].split('\n')) {
+        urlStr = urlStr.trimmed();
+        QUrl url(urlStr);
+        if (url.isValid())
+            trackers << urlStr;
+    }
+    torrent->addTrackers(trackers);
+}
+
+void TorrentsController::editTrackerAction()
+{
+    checkParams({"hash", "origUrl", "newUrl"});
+
+    const QString hash = params()["hash"];
+    const QString origUrl = params()["origUrl"];
+    const QString newUrl = params()["newUrl"];
+
+    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QUrl origTrackerUrl(origUrl);
+    QUrl newTrackerUrl(newUrl);
+    if (origTrackerUrl == newTrackerUrl)
+        return;
+    if (!newTrackerUrl.isValid())
+        throw APIError(APIErrorType::BadParams, "New tracker URL is invalid");
+
+    QList<BitTorrent::TrackerEntry> trackers = torrent->trackers();
```

Comment:
`const` here

Reply:
This can't be const because we modify `trackers`'s members